src/logger/src/logger.rs

```diff
@@ -0,0 +1,611 @@
+    ///
+    /// * `header` - Info about the app that uses the logger.
+    /// * `instance_id` - Unique string identifying this logger session.
+    /// * `log_dest` - Buffer for plain text logs. Needs to implements `Write` and `Send`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// extern crate logger;
+    /// use logger::LOGGER;
+    ///
+    /// use std::io::Cursor;
+    ///
+    /// fn main() {
+    ///     let mut logs = Cursor::new(vec![0; 15]);
+    ///
+    ///     LOGGER.init(
+    ///         "Running Firecracker v.x".to_string(),
+    ///         Box::new(logs),
+    ///     );
+    /// }
+    /// ```
+    pub fn init(&self, header: String, log_dest: Box<dyn Write + Send>) -> Result<()> {
+        self.try_lock(INITIALIZING)?;
+        {
+            let mut g = buf_guard(&self.log_buf);
+
+            *g = Some(log_dest);
+        }
+
+        set_max_level(Level::Trace.to_level_filter());
+
+        STATE.store(INITIALIZED, Ordering::SeqCst);
+        self.log_helper(header, Level::Info);
+
+        Ok(())
+    }
+
+    // In a future PR we'll update the way things are written to the selected destination to avoid
+    // the creation and allocation of unnecessary intermediate Strings. The log_helper method takes
+    // care of the common logic involved in writing regular log messages.
+    fn log_helper(&self, msg: String, msg_level: Level) {
+        if STATE.load(Ordering::Relaxed) == INITIALIZED {
+            if let Some(guard) = buf_guard(&self.log_buf).as_mut() {
+                if write_to_destination(msg, guard).is_err() {
+                    // No reason to log the error to stderr here, just increment the metric.
+                    METRICS.logger.missed_log_count.inc();
+                }
+            } else {
+                METRICS.logger.missed_log_count.inc();
+                panic!("Failed to write to the provided log destination due to poisoned lock");
+            }
+        } else if msg_level <= Level::Warn {
+            eprintln!("{}", msg);
+        } else {
+            println!("{}", msg);
+        }
+    }
+}
+
+/// Implements the "Log" trait from the externally used "log" crate.
+impl Log for Logger {
+    // Test whether the level of the log line should be outputted or not based on the currently
+    // configured level. If the configured level is "warning" but the line is logged through "info!"
+    // marco then it will not get logged.
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() as usize <= self.level.load(Ordering::SeqCst)
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            let msg = format!(
+                "{}{}{}{}",
+                LocalTime::now(),
+                self.create_prefix(&record),
+                MSG_SEPARATOR,
+                record.args()
+            );
+            self.log_helper(msg, record.metadata().level());
+        }
+    }
+
+    // This is currently not used.
```

Comment:
nit: I find it strange that we don't use the trait method to flush the logs. I know that we do it somehow embedded in `self.log_helper`, but still, it would help in terms of clarity if we would have the writing separated from flushing.

Reply:
I agree it would be nice to implement the trait method for flushing, but I don't know if it's worth separating `write` and `flush` operations as long as we will always call them together.